internal/server/pool_test.go

```diff
@@ -53,7 +54,455 @@ func TestPoolSpike(t *testing.T) {
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+
+		be.Equal(t, 0, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+		be.Equal(t, 2, p.Stats().MaxLifetimeClosed)
+	})
+}
+
+func TestMaxLifetimeUpdate(t *testing.T) {
+	cfg := genBasePoolConfig()
+	cfg.MaxOpen = 3
+	cfg.MaxIdle = 3
+	cfg.MaxLifetime = 1 * time.Second
+	cfg.MaxIdleTime = cfg.MaxLifetime * 10
+
+	rt := be.Relaxed(t)
+
+	p, err := NewPool(cfg)
+	be.NilErr(rt, err)
+	defer func() { be.NilErr(rt, p.Close()) }()
+
+	conn0, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	conn1, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+
+	p.ReleaseConn(conn0)
+	p.ReleaseConn(conn1)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	time.Sleep(100 * time.Millisecond)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	p.SetConnMaxLifetime(100 * time.Millisecond)
+
+	time.Sleep(p.maxLifetime + time.Second)
+
+	be.Equal(t, 0, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+	be.Equal(t, 2, p.Stats().MaxLifetimeClosed)
+}
+
+func TestMaxOpenConnBlock(t *testing.T) {
+	cfg := genBasePoolConfig()
+	cfg.MaxOpen = 2
+
+	rt := be.Relaxed(t)
+
+	p, err := NewPool(cfg)
+	be.NilErr(rt, err)
+	defer func() { be.NilErr(rt, p.Close()) }()
+
+	conn0, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	conn1, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	var wg sync.WaitGroup
+	connChan := make(chan *ServerConn, 1)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		sc, err2 := p.AcquireConn()
+		be.NilErr(rt, err2)
+		connChan <- sc
+	}()
+
+	// Wait to let some goroutines wait
+	// and ensure that none can acquire without releasing.
+	time.Sleep(100 * time.Millisecond)
+	be.Equal(t, 2, p.Stats().OpenConnections)
+
+	p.SetMaxOpenConns(3)
+	// allow goroutine to acquire the conn.
+	p.ReleaseConn(conn0)
+	wg.Wait()
+
+	be.Equal(t, 1, p.Stats().WaitCount)
+	be.True(t, p.Stats().WaitDuration >= 100*time.Millisecond)
+
+	// Acquire another one
+	conn0, err = p.AcquireConn()
+	be.NilErr(rt, err)
+
+	be.Equal(t, 3, p.Stats().OpenConnections)
+
+	p.ReleaseConn(<-connChan)
+	p.ReleaseConn(conn1)
+	p.ReleaseConn(conn0)
+}
+
+func TestReloadPool(t *testing.T) {
+	cfg := PoolConfig{
+		SpawnConn: func(ctx context.Context) (Conner, error) {
+			return &connMock{}, nil
+		},
+		MaxOpen:     2,
+		MaxIdle:     5,
+		MaxLifetime: 2 * time.Second,
+		MaxIdleTime: 5 * time.Second,
+	}
+
+	rt := be.Relaxed(t)
+
+	p, err := NewPool(cfg)
+	be.NilErr(rt, err)
+	defer func() { be.NilErr(rt, p.Close()) }()
+
+	cfg.MaxOpen = 4
+	cfg.MaxIdle = 10
+	cfg.MaxLifetime = 10 * time.Second
+	cfg.MaxIdleTime = 8 * time.Second
+
+	p.Reload(cfg)
+
+	be.Equal(t, 4, p.maxOpen)
+	be.Equal(t, 4, p.maxIdle) // maxIdle cannot be > maxOpen
+	be.Equal(t, 10*time.Second, p.maxLifetime)
+	be.Equal(t, 8*time.Second, p.maxIdleTime)
+}
+
+func TestOpenPendingConnWithReleaseExpiry(t *testing.T) {
```

Comment:
I don't understand this test, could you clarify?

Reply:
Yeah good one, I'll add a comment. Basically we are trying to trigger this code path: https://github.com/mattermost/perseus/blob/10f1c2254aa7941ee8f8ae1ae407be6af61b2bae/internal/server/pool.go#L244-L247

and then trigger https://github.com/mattermost/perseus/blob/10f1c2254aa7941ee8f8ae1ae407be6af61b2bae/internal/server/pool.go#L231 because there are open requests pending.

Reply:
Ah, ok, neat! Thanks, a comment will help :)